sql/rpl_master_info_file.hh

```diff
@@ -0,0 +1,622 @@
+      char end;
+      if (sscanf(buf, "%lf%c", &seconds, &end) < 2 || end
+    #else
+      std::from_chars_result result=
+        std::from_chars(buf, &buf[length], seconds, std::chars_format::fixed);
+      if (result.ec != IntIOCache::ERRC_OK || *(result.ptr)
+    #endif
+            != '\n' || seconds < 0 || seconds > SLAVE_MAX_HEARTBEAT_PERIOD)
+        return true;
+      operator=(static_cast<uint32_t>(seconds / 1000));
+      return false;
+    }
+    /**
+      This method is engineered (that is, hard-coded) to take
+      full advantage of the non-negative `DECIMAL(10,3)` format.
+    */
+    void save_to(IO_CACHE *file) override {
+      char buf[IntIOCache::BUF_SIZE<uint32_t>];
+      std::to_chars_result result=
+        std::to_chars(buf, &buf[sizeof(buf)], operator uint32_t());
+      DBUG_ASSERT(result.ec == IntIOCache::ERRC_OK);
+      ptrdiff_t size= result.ptr - buf;
+      if (size > 3) // decimal seconds has ones digit or more
+      {
+        my_b_write(file, (const uchar *)buf, size - 3);
+        my_b_write_byte(file, '.');
+        my_b_write(file, (const uchar *)(&result.ptr[-3]), 3);
+      }
+      else
+      {
+        my_b_write_byte(file, '0');
+        my_b_write_byte(file, '.');
+        for (ptrdiff_t zeroes= size; zeroes < 3; ++zeroes)
+          my_b_write_byte(file, '0');
+        my_b_write(file, (const uchar *)buf, size);
+      }
+    }
+  }
+  /// `Slave_heartbeat_period` of SHOW ALL SLAVES STATUS
+  master_heartbeat_period;
+
+  /// }@
+
+
+  inline static const std::initializer_list<mem_fn> FIELDS_LIST= {
+    &MasterInfoFile::master_log_file,
+    &MasterInfoFile::master_log_pos,
+    &MasterInfoFile::master_host,
+    &MasterInfoFile::master_user,
+    &MasterInfoFile::master_password,
+    &MasterInfoFile::master_port,
+    &MasterInfoFile::master_connect_retry,
+    &MasterInfoFile::master_ssl,
+    &MasterInfoFile::master_ssl_ca,
+    &MasterInfoFile::master_ssl_capath,
+    &MasterInfoFile::master_ssl_cert,
+    &MasterInfoFile::master_ssl_cipher,
+    &MasterInfoFile::master_ssl_key,
+    &MasterInfoFile::master_ssl_verify_server_cert,
+    &MasterInfoFile::master_heartbeat_period,
+    nullptr, // &MasterInfoFile::master_bind, // MDEV-19248
+    &MasterInfoFile::ignore_server_ids,
+    nullptr, // MySQL field `master_uuid`, which MariaDB ignores.
+    &MasterInfoFile::master_retry_count,
+    &MasterInfoFile::master_ssl_crl,
+    &MasterInfoFile::master_ssl_crlpath
+  };
+
+  /**
+    Guard agaist extra left-overs at the end of file in case a later update
+    causes the effective content to shrink compared to earlier contents
+  */
+  static constexpr const char END_MARKER[]= "END_MARKER";
+  /// An iterable for the `key=value` section of `@@master_info_file`
+  // C++ default allocator to match that `mysql_execute_command()` uses `new`
+  inline static const std::unordered_map<std::string_view, mem_fn> FIELDS_MAP= {
```

Comment:
Why unordered? As the `FIELDS_MAP` is iterated over when saving the KV pairs, wouldn't that make it non-deterministic? The fields should be written in the same order as pre-MDEV-37530 so users can downgrade without breaking anything.

Reply:
The pre-MDEV-37530 code checks keys, too, so the order is not important.
I can see that the predecessor who added this key-value section in the first place had order independence in mind.

(That code is a chain of `if (got_eq && !seen_X && !strcmp(buf, "X"))`, however, which will look ridiculous under my (backward-compatible!) plan of stuffing the `DEFAULT` choices in that section, not to mention performance comparisons with hash lookups.)